logstash-core/lib/logstash/agent.rb

```diff
@@ -11,32 +12,78 @@
 LogStash::Environment.load_locale!
 
 class LogStash::Agent
+  attr_reader :logger, :state
 
-  attr_writer :logger
-  attr_reader :node_name
-
-  def initialize(options = {})
+  def initialize(params)
+    @logger = params[:logger]
     @pipelines = {}
      
-    @node_name = options[:node_name] || Socket.gethostname
+    @node_name = params[:node_name] || Socket.gethostname
+    @pipeline_threads = {}
+    @config_loader = LogStash::Config::Loader.new(@logger, false)
+    @config_string = params[:config_string]
+    @config_path = params[:config_path]
+    @auto_reload = params[:auto_reload]
+    @pipeline_settings = params[:pipeline_settings]
+    @reload_interval = params[:reload_interval] || 3 # seconds
+    @upgrade_mutex = Mutex.new
+    # @state represents the necessary information for the agent's lifecycle.
+    # this base implementation uses the configuration string itself as state.
+    # if fetch_state returns a different string, it is regarded as new state
+    # and upgrade_state will take care of stopping the previous pipeline and
+    # starting a new one
+    # a subclass of this base Agent could use more complex state objects such
+    # as a hash map containing versioning, update_at timestamp, etc.
+    @state = clean_state
   end
 
   def execute
-    # Make SIGINT/SIGTERM shutdown the pipeline.
-    sigint_id = trap_sigint()
-    sigterm_id = trap_sigterm()
+    @thread = Thread.current
```

Comment:
+1. Also maybe adding an api to Stud that makes it more obvious, instead of an ivar setting?